exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/common/HashTableTemplate.java

```diff
@@ -544,86 +572,69 @@ private static int roundUpToPowerOf2(int number) {
+      lastEntryIdxWithinBatch = currentIdxHolder.value & BATCH_MASK;
 
-      if (EXTRA_DEBUG) {
-        logger.debug("Empty bucket index = {}. incomingRowIdx = {}; inserting new entry at currentIdx = {}.", i,
-            incomingRowIdx, currentIdx);
+      if (lastEntryBatch.isKeyMatch(incomingRowIdx, currentIdxHolder, false)) {
+        htIdxHolder.value = currentIdxHolder.value;
+        return PutStatus.KEY_PRESENT;
       }
-
-      insertEntry(incomingRowIdx, currentIdx, hash, lastEntryBatch, lastEntryIdxWithinBatch);
-      // update the start index array
-      startIndices.getMutator().setSafe(getBucketIndex(hash, numBuckets()), currentIdx);
-      htIdxHolder.value = currentIdx;
-      return PutStatus.KEY_ADDED;
     }
 
-    currentIdx = startIdx;
-    boolean found = false;
-
-    bh = batchHolders.get((currentIdx >>> 16) & BATCH_MASK);
-    currentIdxHolder.value = currentIdx;
-
-    // if startIdx is non-empty, follow the hash chain links until we find a matching
-    // key or reach the end of the chain
-    while (true) {
-      currentIdxWithinBatch = currentIdxHolder.value & BATCH_MASK;
+    // no match was found, so insert a new entry
+    currentIdx = freeIndex++;
+    boolean addedBatch = addBatchIfNeeded(currentIdx);
 
-      if (bh.isKeyMatch(incomingRowIdx, currentIdxHolder, false)) {
-        htIdxHolder.value = currentIdxHolder.value;
-        found = true;
-        break;
-      } else if (currentIdxHolder.value == EMPTY_SLOT) {
-        lastEntryBatch = bh;
-        lastEntryIdxWithinBatch = currentIdxWithinBatch;
-        break;
-      } else {
-        bh = batchHolders.get((currentIdxHolder.value >>> 16) & HashTable.BATCH_MASK);
-        lastEntryBatch = bh;
-      }
+    if (EXTRA_DEBUG) {
+      logger.debug("No match was found for incomingRowIdx = {}; inserting new entry at currentIdx = {}.", incomingRowIdx, currentIdx);
     }
 
-    if (!found) {
-      // no match was found, so insert a new entry
-      currentIdx = freeIndex++;
-      addBatchIfNeeded(currentIdx);
+    insertEntry(incomingRowIdx, currentIdx, hashCode, lastEntryBatch, lastEntryIdxWithinBatch);
 
-      if (EXTRA_DEBUG) {
-        logger.debug("No match was found for incomingRowIdx = {}; inserting new entry at currentIdx = {}.", incomingRowIdx, currentIdx);
-      }
-
-      insertEntry(incomingRowIdx, currentIdx, hash, lastEntryBatch, lastEntryIdxWithinBatch);
-      htIdxHolder.value = currentIdx;
-      return PutStatus.KEY_ADDED;
+    // if there was no hash chain at this bucket, need to update the start index array
+    if (startIdx == EMPTY_SLOT) {
+      startIndices.getMutator().setSafe(getBucketIndex(hashCode, numBuckets()), currentIdx);
     }
-
-    return found ? PutStatus.KEY_PRESENT : PutStatus.KEY_ADDED;
+    htIdxHolder.value = currentIdx;
+    return  addedBatch ? PutStatus.NEW_BATCH_ADDED :
+        ( freeIndex + 1 > batchHolders.size() * BATCH_SIZE ) ?
```

Comment:
The former: Last key in a batch. This information is needed at the Hash Aggr to initiate a check for memory pressure (could calculate this information, but looks cleaner/simpler to get that as a special status from the hash table). Hopefully we'd never have to deal with batches of size 1 (row), as it would conflict - both "last row" and "new batch" at the same insertion.